### ISSUE TITLE
sneaking on a block with a hole kills me

### ISSUE THREAD

**Issue by kingoscargames** [minetest/minetest#5970]
When i sneak on this block and then stop sneaking or move a little,
![screenshot_20170612_104300](https://user-images.githubusercontent.com/29333817/27029648-8d3fda52-4f69-11e7-8a3f-3fbe40922b4a.png)
it hurts my character,
![screenshot_20170612_104302](https://user-images.githubusercontent.com/29333817/27029649-8d41f71a-4f69-11e7-807a-461d2f454e7c.png)
And if i hold it long enough i die.
![screenshot_20170612_104307](https://user-images.githubusercontent.com/29333817/27029647-8d3c5e86-4f69-11e7-99b0-fe37f5b80ec4.png)
This only happened after installing the new release.
Using this subgame:https://forum.minetest.net/viewtopic.php?f=50&t=15925&hilit=ocean&sid=a1923f8df2a41ed099d47f4dacc4e066

but when i use this mod https://forum.minetest.net/viewtopic.php?t=17734 it doesnt hurt me.

**Comment by kingoscargames** [comment]
I just tested again it doesn't happen in the previous release.

**Comment by kingoscargames** [comment]
Its a bug with sneaking because when i re-enable the old code (with this mod:https://forum.minetest.net/viewtopic.php?t=17734),
it doesn't happen anymore.

**Comment by kingoscargames** [comment]
I copied the mod to minetest_game 0.4.15 (still use it in the newer version) and it still happens.
![screenshot_20170612_124657](https://user-images.githubusercontent.com/29333817/27030741-b5268e40-4f6d-11e7-97cf-7bfa3f49eb28.png)

**Comment by kingoscargames** [comment]
well another reason to use the mod then.

**Comment by paramat** [comment]
![screenshot_20170325_002845](https://cloud.githubusercontent.com/assets/3686677/24317712/169e99a2-10f2-11e7-9c81-ea72c8633c54.png)

I can confirm when using new move code (default in 0.4.16). If i sneak and move forwards over the empty part of this node (a rotated stair) the player vibrates up and down and is then killed after 2-3s, as if falling.
Seems to me the code thinks the player is falling when sneaking over a void in a node.

**Comment by tenplus1** [comment]
Could this be related in any way : https://github.com/minetest/minetest/issues/5966

**Comment by SmallJoker** [comment]
Indeed, this came up with the new sneak code.
```C++
// Because we keep the player collision box on the node, limiting
// position.Y is not necessary but useful to prevent players from
// being inside a node if sneaking on e.g. the lower part of a stair
if (!m_sneak_ladder_detected) {
	position.Y = MYMAX(position.Y, bmax.Y);
} else {
```
Sauce: https://github.com/minetest/minetest/blob/master/src/localplayer.cpp#L330
The player in this situation stands on the most upper nodebox(-es) of the current node, which is in the above shown nodes always a node of 1x1x1 m (maximal side lengths). But at the same time, the collision detection accelerates the player because at his place, there's actually no node where the player could stand on.
Due the code shown here, the player gets teleported back to the most upper Y position of the node, but doesn't reset the velocity. At a point, the maximal possible velocity is reached and the player collides with the node below.

This problem could get solved by adding `m_speed.Y = MYMAX(m_speed.Y, 0);` into this case aswell, but new side-effects are almost guaranteed. The sneak code is a minefield.

PS: Got a déjà vu - I think there's already an issue about this specific part of the code.

**Comment by sfan5** [comment]
Well this is definitely a bug, the "new" sneak code should keep the player collisionbox ontop of the node but apparently fails to do so.

**Comment by paramat** [comment]
As far as i remember the initial new sneak code didn't do this, i think this started with a commit by SmallJoker?
With the original code if i sneaked on the rotated stair shown and moved towards the cutout section i would fall off the node.

**Comment by SmallJoker** [comment]
Tested on current HEAD, bug fixed by #5533.
